### PR TITLE
Support CentOS based distros

### PR DESCRIPTION
**Pull request by lindenb1**
# Package Comparisson

| Ubuntu | CentOS |
| --- | --- |
| ca-certificates | ca-certificates |
| clevis | clevis |
| clevis-dracut                | clevis-dracut                |
| clevis-initramfs             | UNKNOWN/UNAVAILABLE          |
| clevis-luks                  | clevis-luks                  |
| git                          | git                          |
| docker-compose               | docker-compose-plugin        |
| docker-compose-plugin        | docker-compose-plugin        |
| ufw (Not part of rocky defualt image, needs to be installed) | ufw  |
| ipmitool                     | ipmitool                     |
| kubectl                     | kubectl (https://prod-cdn.packages.k8s.io/repositories/isv:/kubernetes:/core:/stable:/v1.28/rpm/repodata/repomd.xml.key) |
| lynis                        | lynis                        |
| amd64-microcode              | linux-firmware               |
| intel-microcode              | linux-firmware               |
| update-motd                  | UNKNOWN/UNAVAILABLE          |
| netplan.io                   | netplan.io                   |
| ethtool                      | ethtool                      |
| jq                           | jq                           |
| rsyslog                      | rsyslog                      |
| podman                       | podman                       |
| openresolv                   | UNKNOWN/UNAVAILABLE          |
| resolvconf                   | UNKNOWN/UNAVAILABLE          |
| runc                         | runc                         |
| sosreport | sos -> also uses different plugins! |

# Kubectl Repo Setup

sudo cat <<EOF > /etc/yum.repos.d/kubernetes.repo
[kubernetes]
name=Kubernetes
baseurl=https://pkgs.k8s.io/core:/stable:/v1.28/rpm/
enabled=1
gpgcheck=1
gpgkey=https://pkgs.k8s.io/core:/stable:/v1.28/rpm/repodata/repomd.xml.key
EOF

# Trivy Repo Setup

sudo tee /etc/yum.repos.d/trivy.repo <<EOL
[trivy]
name=Trivy Repository
baseurl=https://aquasecurity.github.io/trivy-repo/rpm/releases/$basearch/
gpgcheck=1
enabled=1
gpgkey=https://aquasecurity.github.io/trivy-repo/rpm/public.key
EOL

# Implementation Check
cf. https://github.com/osism/issues/issues/844

- [x] ansible-collection-commons-molecule-certificates
- [x] ansible-collection-commons-molecule-cleanup
- [x] ansible-collection-commons-molecule-configfs
- [x] ansible-collection-commons-molecule-configuration
- [x] ansible-collection-commons-molecule-docker_compose
- [ ] ansible-collection-commons-molecule-docker_login
- [x] ansible-collection-commons-molecule-facts
- [x] ansible-collection-commons-molecule-firewall
- [x] ansible-collection-commons-molecule-hostname
- [x] ansible-collection-commons-molecule-hosts
- [x] ansible-collection-commons-molecule-ipmitool
- [x] ansible-collection-commons-molecule-kernel_modules
- [x] ansible-collection-commons-molecule-known_hosts
- [x] ansible-collection-commons-molecule-kompose
- [x] ansible-collection-commons-molecule-kubectl
- [x] ansible-collection-commons-molecule-limits
- [x] ansible-collection-commons-molecule-lynis
- [x] ansible-collection-commons-molecule-microcode
- [x] ansible-collection-commons-molecule-motd
- [ ] ansible-collections-commons-molecule-network
- [x] ansible-collection-commons-molecule-operator
- [x] ansible-collection-commons-molecule-packages
- [x] ansible-collection-commons-molecule-podman
- [x] ansible-collection-commons-molecule-proxy
- [ ] ansible-collection-commons-molecule-repository
- [ ] ansible-collection-commons-molecule-resolvconf
- [x] ansible-collection-commons-molecule-runc
- [x] ansible-collection-commons-molecule-services
- [x] ansible-collection-commons-molecule-sosreport
- [x] ansible-collection-commons-molecule-sshconfig
- [x] ansible-collection-commons-molecule-state
- [x] ansible-collection-commons-molecule-sysctl
- [x] ansible-collection-commons-molecule-systohc
- [x] ansible-collection-commons-molecule-timezone
- [x] ansible-collection-commons-molecule-trivy
- [ ] ansible-collection-commons-molecule-user :new:

### CODE REVIEW

**Comment by sbstnnmnn** [comment]
- [x] ansible-collection-commons-molecule-certificates
- [x] ansible-collection-commons-molecule-cleanup
- [x] ansible-collection-commons-molecule-configfs
- [ ] ansible-collection-commons-molecule-configuration
- [ ] ansible-collection-commons-molecule-docker_compose
- [ ] ansible-collection-commons-molecule-docker_login
- [x] ansible-collection-commons-molecule-facts
- [ ] ansible-collection-commons-molecule-firewall
- [x] ansible-collection-commons-molecule-hostname
- [x] ansible-collection-commons-molecule-hosts
- [ ] ansible-collection-commons-molecule-ipmitool
- [x] ansible-collection-commons-molecule-kernel_modules
- [x] ansible-collection-commons-molecule-known_hosts
- [ ] ansible-collection-commons-molecule-kompose
- [ ] ansible-collection-commons-molecule-kubectl
- [x] ansible-collection-commons-molecule-limits
- [ ] ansible-collection-commons-molecule-lynis
- [ ] ansible-collection-commons-molecule-microcode
- [ ] ansible-collection-commons-molecule-motd
- [ ] ansible-collection-commons-molecule-operator
- [x] ansible-collection-commons-molecule-packages
- [ ] ansible-collection-commons-molecule-podman
- [x] ansible-collection-commons-molecule-proxy
- [ ] ansible-collection-commons-molecule-repository
- [ ] ansible-collection-commons-molecule-resolvconf
- [ ] ansible-collection-commons-molecule-runc
- [x] ansible-collection-commons-molecule-services
- [ ] ansible-collection-commons-molecule-sosreport
- [x] ansible-collection-commons-molecule-sshconfig
- [x] ansible-collection-commons-molecule-state
- [x] ansible-collection-commons-molecule-sysctl
- [x] ansible-collection-commons-molecule-systohc
- [ ] ansible-collection-commons-molecule-timezone
- [ ] ansible-collection-commons-molecule-trivy

File: molecule/delegated/tests/kubectl/redhat.py [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+    check_configure_repository(host)
+
+    all_keys = host.run(
+        "rpm -qa gpg-pubkey* --qf '%{NAME}-%{VERSION}-%{RELEASE} --> %{SUMMARY}\n%{DESCRIPTION}\n\n'"
+    ).stdout
+    installed_key = get_centos_repo_key(
+        host,
+        "isv:kubernetes OBS Project <isv:kubernetes@build.opensuse.org> public key",
+    )
+    assert installed_key in all_keys
```
Comment (gndrmnn) [inline molecule/delegated/tests/kubectl/redhat.py:36]: Redundant check, we just want to know if get_centos_repo_key is the same as in the file

File: molecule/delegated/tests/motd/debian.py [hunk 2]
```diff
@@ -14,3 +14,5 @@ def test_pkg(host):
     check_ansible_os_family(host)
 
     assert not host.package("update-motd").is_installed
+
+    # There isn't a direct equivalent to the "update-motd" package on CentOS.
```
Comment (gndrmnn) [inline molecule/delegated/tests/motd/debian.py:18]: Is this really needed? There is no redhat.py which should be conclusive enough. Adding a comment here about centos seems out of place (but i know why you put it here of course)

File: molecule/delegated/tests/repository/redhat.py [hunk 3]
```diff
@@ -0,0 +1,13 @@
...
+    get_ansible,
+    get_variable,
+)
+
+testinfra_runner, testinfra_hosts = get_ansible()
+
+
+def check_ansible_os_family(host):
+    if get_variable(host, "ansible_os_family", True) != "RedHat":
+        pytest.skip("ansible_os_family mismatch")
```
Comment (gndrmnn) [inline molecule/delegated/tests/repository/redhat.py:13]: ???

File: molecule/delegated/tests/resolvconf/redhat.py [hunk 4]
```diff
@@ -0,0 +1,13 @@
...
+from ..util.util import (
+    get_ansible,
+    get_variable,
+)
+
+testinfra_runner, testinfra_hosts = get_ansible()
+
+
+def check_ansible_os_family(host):
+    if get_variable(host, "ansible_os_family", True) != "RedHat":
```
Comment (gndrmnn) [inline molecule/delegated/tests/resolvconf/redhat.py:12]: ???

File: molecule/delegated/tests/trivy/redhat.py [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+
+
+def test_package(host):
+    """Check if the packages are installed."""
+    check_ansible_os_family(host)
+
+    if get_variable(host, "trivy_configure_repository"):
+        pkg_name = "yum-utils"
+        pkg = host.package(pkg_name)
+        assert pkg.is_installed
```
Comment (gndrmnn) [inline molecule/delegated/tests/trivy/redhat.py:26]: We dont need to test for the packages that were needed to install the actual packages

File: molecule/delegated/tests/util/util.py [hunk 6]
```diff
@@ -89,6 +89,19 @@ def get_from_url(url, binary=False):
     return content
 
 
+def get_centos_repo_key(host, summary):
```
Comment (gndrmnn) [inline molecule/delegated/tests/util/util.py:92]: I would argue it would be good to add Robins UserAgent spoof from collection-services

File: roles/certificates/defaults/main.yml [hunk 7]
```diff
@@ -5,7 +5,8 @@
 #       -----BEGIN CERTIFICATE-----
 #       [...]
 #       -----END CERTIFICATE-----
+
 certificates_ca: []
-certificates_ca_path: /usr/local/share/ca-certificates
 certificates_ca_package_name: ca-certificates
-certificates_ca_update_command: /usr/sbin/update-ca-certificates
+certificates_ca_path: ""
+certificates_ca_update_command: ""
```
Comment (gndrmnn) [inline roles/certificates/defaults/main.yml:12]: Can be removed if they are defined in ``/vars``

File: roles/certificates/tasks/install-Debian.yml [hunk 8]
```diff
@@ -1,7 +1,7 @@
 ---
-- name: Install ca-certificates package
+- name: Install ca-certificates package on Debian
   become: true
   ansible.builtin.apt:
     name: "{{ certificates_ca_package_name }}"
     state: present
-    lock_timeout: "{{ apt_lock_timeout | default(300) }}"
+    lock_timeout: "{{ lock_timeout | default(300) }}"
```
Comment (gndrmnn) [inline roles/certificates/tasks/install-Debian.yml:7]: Why that change?

File: roles/configuration/tasks/git.yml [hunk 9]
```diff
@@ -23,12 +23,21 @@
...
-    lock_timeout: "{{ apt_lock_timeout | default(300) }}"
+    lock_timeout: "{{ lock_timeout | default(300) }}"
+  when: "'Debian' in ansible_os_family"
+
+- name: Git - install required packages on RedHat
+  become: true
+  ansible.builtin.dnf:
+    name: "{{ configuration_git_package_name }}"
+    state: present
+    lock_timeout: "{{ lock_timeout | default(300) }}"
```
Comment (gndrmnn) [inline roles/configuration/tasks/git.yml:39]: Does the lock_timeout actually exist? I dont get the change here

File: roles/lynis/tasks/install-Debian.yml [hunk 10]
```diff
@@ -2,7 +2,7 @@
 - name: Remove old architecture-dependent repository
   become: true
   ansible.builtin.apt_repository:
-    repo: "deb [ arch=amd64 ] https://packages.cisofy.com/community/lynis/deb/ stable main"
+    repo: "deb [arch=amd64] https://packages.cisofy.com/community/lynis/deb/ stable main"
```
Comment (gndrmnn) [inline roles/lynis/tasks/install-Debian.yml:5]: why the change?

File: roles/sosreport/defaults/main.yml [hunk 11]
```diff
@@ -2,36 +2,10 @@
 sosreport_unarchive: false
 sosreport_tmpdir: /tmp/sosreport
 sosreport_cmd: sosreport
-sosreport_required_packages:
-  - sosreport
+sosreport_required_packages: ""
```
Comment (gndrmnn) [inline roles/sosreport/defaults/main.yml:5]: not sure if we need this if this is present in /vars

File: roles/sosreport/defaults/main.yml [hunk 12]
```diff
@@ -2,36 +2,10 @@
...
-  - python
-  - services
-  - ssh
-  - system
-  - systemd
-  - ubuntu
-  - udev
-  - usb
-  - xfs
+sosreport_plugins: ""
```
Comment (gndrmnn) [inline roles/sosreport/defaults/main.yml:9]: not sure if we need this if this is present in /vars

File: roles/timezone/tasks/main.yml [hunk 13]
```diff
@@ -14,3 +14,16 @@
...
+  ansible.builtin.file:
+    path: /etc/adjtime
+    state: touch
+    mode: '0644'
+
+- name: Ensure UTC in /etc/adjtime
+  become: true
+  ansible.builtin.lineinfile:
+    path: /etc/adjtime
+    line: "{{ timezone_hwclock }}"
```
Comment (gndrmnn) [inline roles/timezone/tasks/main.yml:29]: Why was this added?

File: molecule/delegated/tests/certificates/redhat.py [hunk 14]
```diff
@@ -0,0 +1,20 @@
+import pytest
```
[the remark this answers is not in the document]
Comment (sbstnnmnn) [reply molecule/delegated/tests/certificates/redhat.py]: done

File: molecule/delegated/tests/motd/debian.py (hunk 2 above)
Comment (sbstnnmnn) [reply molecule/delegated/tests/motd/debian.py:18]: deleted

File: molecule/delegated/tests/resolvconf/debian.py [hunk 15]
```diff
@@ -1,4 +1,17 @@
...
+from ..util.util import (
+    get_ansible,
+    get_variable,
+    jinja_list_concat,
+)
+
+testinfra_runner, testinfra_hosts = get_ansible()
+
+
+def check_ansible_os_family(host):
```
[the remark this answers is not in the document]
Comment (sbstnnmnn) [reply molecule/delegated/tests/resolvconf/debian.py:12]: Added to some functions. 
Was preparatory work for RHEL implementation of role and test.

File: molecule/delegated/tests/resolvconf/redhat.py (hunk 4 above)
Comment (sbstnnmnn) [reply molecule/delegated/tests/resolvconf/redhat.py:12]: deleted for the moment

File: molecule/delegated/tests/trivy/redhat.py (hunk 5 above)
Comment (sbstnnmnn) [reply molecule/delegated/tests/trivy/redhat.py:26]: ok

File: roles/certificates/defaults/main.yml (hunk 7 above)
Comment (sbstnnmnn) [reply roles/certificates/defaults/main.yml:12]: They are, absolutely right.

File: roles/certificates/tasks/install-Debian.yml (hunk 8 above)
Comment (sbstnnmnn) [reply roles/certificates/tasks/install-Debian.yml:7]: changed back

File: roles/lynis/tasks/install-Debian.yml (hunk 10 above)
Comment (sbstnnmnn) [reply roles/lynis/tasks/install-Debian.yml:5]: makes no different

File: roles/sosreport/defaults/main.yml (hunk 11 above)
Comment (sbstnnmnn) [reply roles/sosreport/defaults/main.yml:5]: probably not

File: roles/sosreport/defaults/main.yml (hunk 12 above)
Comment (sbstnnmnn) [reply roles/sosreport/defaults/main.yml:9]: probably not

File: roles/timezone/tasks/main.yml (hunk 13 above)
Comment (sbstnnmnn) [reply roles/timezone/tasks/main.yml]: Only for CentOS needed. changed tasks this way: 

```
- name: RedHat specific tasks
  become: true
  block:
    - name: Create /etc/adjtime file if not exists
      ansible.builtin.file:
        path: /etc/adjtime
        state: touch
        mode: '0644'
    - name: Ensure UTC in /etc/adjtime
      ansible.builtin.lineinfile:
        path: /etc/adjtime
        line: "{{ timezone_hwclock }}"
  when: "'RedHat' in ansible_os_family"
```

File: molecule/delegated/tests/util/util.py (hunk 6 above)
Comment (sbstnnmnn) [reply molecule/delegated/tests/util/util.py:92]: Refactored `get_from_url` function in `util.py`:
```
def get_from_url(url, binary=False):
    # Create a request object with a faked User-Agent header.
    # Some websites like https://pkg.osquery.io/rpm/GPG need this, otherwise they will return http 403 forbidden.
    req = urllib.request.Request(
        url, headers={"User-Agent": "Mozilla/5.0 (Linux x86_64) Chrome/103.0.0.0"}
    )
    # Open the URL with the custom request object
    resource = urllib.request.urlopen(req)

    if not binary:
        encoding = resource.headers.get_content_charset()
        if encoding is None:
            encoding = "utf-8"
        content = resource.read().decode(encoding)
    else:
        content = resource.read()

    return content
```

File: roles/sosreport/defaults/main.yml (hunk 11 above)
Comment (sbstnnmnn) [reply roles/sosreport/defaults/main.yml]: cf. molecule/delegated/tests/sosreport.py

**Review by gndrmnn — APPROVED** [review]
